README.rst
Description
===========

.. image:: https://secure.travis-ci.org/rvanlaar/djangorecipe.png?branch=master
   :target: http://travis-ci.org/rvanlaar/djangorecipe/

This buildout recipe can be used to create a setup for Django. It will
automatically download Django and install it in the buildout's
sandbox.

You can see an example of how to use the recipe below::

  [buildout]
  parts = satchmo django
  eggs =
    ipython
  versions = versions

  [versions]
  django = 1.2.5


  [satchmo]
  recipe = gocept.download
  url = http://www.satchmoproject.com/snapshots/satchmo-0.6.tar.gz
  md5sum = 659a4845c1c731be5cfe29bfcc5d14b1

  [django]
  recipe = djangorecipe
  settings = development
  eggs = ${buildout:eggs}
  extra-paths =
    ${satchmo:location}
  project = dummyshop


Supported options
=================

The recipe supports the following options.

project
  This option sets the name for your project. The recipe will create a
  basic structure if the project is not already there.

projectegg
  Use this instead of the project option when you want to use an egg
  as the project. This disables the generation of the project
  structure.

python
  This option can be used to specify a specific Python version which can be a
  different version from the one used to run the buildout.

settings
  You can set the name of the settings file which is to be used with
  this option. This is useful if you want to have a different
  production setup from your development setup. It defaults to
  `development`.

  If specified module name does not contains dots, it is expected, that
  settings module is submodule of you project. For example if you project or
  projectegg name is `myproject`, then by default settings module will be
  `myproject.development`.

  If you want to specify absolute module name outside project module, you can
  prepend it with single dot, for example: `.mysettings`, dot will be stripped
  and settings will be searched in absolute `mysettings` module.

extra-paths
  All paths specified here will be used to extend the default Python
  path for the `bin/*` scripts.

pth-files
  Adds paths found from a site `.pth` file to the extra-paths.
  Useful for things like Pinax which maintains its own external_libs dir.

control-script
  The name of the script created in the bin folder. This script is the
  equivalent of the `manage.py` Django normally creates. By default it
  uses the name of the section (the part between the `[ ]`).

initialization
  Specify some Python initialization code to be inserted into the
  `control-script`. This is very limited. In particular, be aware that
  leading whitespace is stripped from the code given.

wsgi
  An extra script is generated in the bin folder when this is set to
  `true`. This can be used with mod_wsgi to deploy the project. The
  name of the script is `control-script.wsgi`.

wsgilog
  In case the WSGI server you're using does not allow printing to stdout,
  you can set this variable to a filesystem path - all stdout/stderr data
  is redirected to the log instead of printed

fcgi
  Like `wsgi` this creates an extra script within the bin folder. This
  script can be used with an FCGI deployment.

test
  If you want a script in the bin folder to run all the tests for a
  specific set of apps this is the option you would use. Set this to
  the list of app labels which you want to be tested.

testrunner
  This is the name of the testrunner which will be created. It
  defaults to `test`.

All following options only have effect when the project specified by
the project option has not been created already.

urlconf
  You can set this to a specific url conf. It will use project.urls by
  default.

secret
  The secret to use for the `settings.py`, it generates a random
  string by default.


FCGI specific settings
======================

Options for FCGI can be set within a settings file (`settings.py`). The options
is `FCGI_OPTIONS`. It should be set to a dictionary. The part below is an
example::

  FCGI_OPTIONS = {
      'method': 'threaded',
  }


Another example
===============

The next example shows you how to use some more of the options::

  [buildout]
  parts = django extras
  eggs =
    hashlib

  [extras]
  recipe = iw.recipe.subversion
  urls =
    http://django-command-extensions.googlecode.com/svn/trunk/ django-command-extensions
    http://django-mptt.googlecode.com/svn/trunk/ django-mptt

  [django]
  recipe = djangorecipe
  settings = development
  project = exampleproject
  wsgi = true
  eggs =
    ${buildout:eggs}
  test =
    someapp
    anotherapp

Example using .pth files
========================

Pinax uses a .pth file to add a bunch of libraries to its path; we can
specify it's directory to get the libraries it specified added to our
path::

  [buildout]
  parts	= PIL
	  svncode
	  myproject
  versions=versions

  [versions]
  django	= 1.3

  [PIL]
  recipe	= zc.recipe.egg:custom
  egg		= PIL
  find-links	= http://dist.repoze.org/

  [svncode]
  recipe	= iw.recipe.subversion
  urls		= http://svn.pinaxproject.com/pinax/tags/0.5.1rc1	pinax

  [myproject]
  recipe	= djangorecipe
  eggs		=
    PIL
  project	= myproject
  settings	= settings
  extra-paths	= ${buildout:directory}/myproject/apps
		  ${svncode:location}/pinax/apps/external_apps
		  ${svncode:location}/pinax/apps/local_apps
  pth-files	= ${svncode:location}/pinax/libs/external_libs
  wsgi		= true

Above, we use stock Pinax for pth-files and extra-paths paths for
apps, and our own project for the path that will be found first in the
list.  Note that we expect our project to be checked out (e.g., by
svn:external) directly under this directory in to 'myproject'.

Example with a different Python version
=======================================

To use a different Python version from the one that ran buildout in the
generated script use something like::

  [buildout]
  parts	= myproject

  [special-python]
  executable = /some/special/python

  [myproject]
  recipe	= djangorecipe
  project	= myproject
  python	= special-python

Example with a Django version from a repository
===============================================

If you want to use a specific Django version from a source 
repository you could use mr.developer: http://pypi.python.org/pypi/mr.developer
Here is an example for using the Django development version::

  [buildout]
  parts = django
  extensions = mr.developer
  auto-checkout = *

  [sources]
  django = git https://github.com/django/django.git

  [django]
  recipe = djangorecipe
  settings = settings
  project = project

Example configuration for mod_wsgi
==================================

If you want to deploy a project using mod_wsgi you could use this
example as a starting point::

  <Directory /path/to/buildout>
         Order deny,allow
         Allow from all
  </Directory>
  <VirtualHost 1.2.3.4:80>
         ServerName      my.rocking.server
         CustomLog       /var/log/apache2/my.rocking.server/access.log combined
         ErrorLog        /var/log/apache2/my.rocking.server/error.log
         WSGIScriptAlias / /path/to/buildout/bin/django.wsgi
  </VirtualHost>

Generating a control script for PyDev
=====================================

Running Django with auto-reload in PyDev requires adding a small snippet
of code::

  import pydevd
  pydevd.patch_django_autoreload(patch_remote_debugger=False, patch_show_console=True)

just before the `if __name__ == "__main__":` in the `manage.py` module
(or in this case the control script that is generated). This example
buildout generates two control scripts: one for command-line usage and
one for PyDev, with the required snippet, using the recipe's
`initialization` option::

  [buildout]
  parts = django pydev
  eggs =
    mock

  [django]
  recipe = djangorecipe
  eggs = ${buildout:eggs}
  project = dummyshop

  [pydev]
  <= django
  initialization =
    import pydevd
    pydevd.patch_django_autoreload(patch_remote_debugger=False, patch_show_console=True)


CHANGES.rst
Changes
=======

1.4 (unreleased)
----------------

- Added (pydev) initialization support. Thanks to anshumanb. (Closes #58).

- Added possibility to specify settings module outside of project module.
  (Closes #52).


1.3 (2012-09-07)
----------------

- Removed deprecation warning in Django 1.4. Fixes #49, thanks Shagi.

- Added documentation for use with mr.developer. Thanks shagi (closes issue #45)

- Added Travis support.


1.2.1 (2012-05-15)
------------------

- Fixed broken 1.2 release (missing ``*.rst`` files due to a recent txt-to-rst
  rename action).


1.2 (2012-05-14)
----------------

- Removed location path from recipe. Thanks bleskes (fixes issue #50).

1.1.2
-----

- Added correct url to the deprecation warning

1.1.1
-----

- Fixed Python3 Trove classifiers

1.1
---

- Support python3.
- Changed buildout and the tests to run the tests under nose.
- Removed some old pre 0.99 unittests that dealt with download support.

1.0
---

- Stable release with a real 1.0 version.
- Made djangorecipe more pep08 compliant.

0.99
----

- Djangorecipe now depends on Django. The use of the `version =` statement
  is deprecated. Specify the django version in the
  `[versions]` section. Install django via mr.developer if you need to use
  an svn/git/hg repository. For other uses
  Versionpin djangorecipe to 0.23.1 if you don't want to upgrade.
  Thanks to Reinout van Rees for help with this release.

- Removed subversion and download support.

0.23.1
------

- Added a missing 'import os'

0.23
----

- Support for settings/urls boilerplate for django 1.2 and django 1.3.
  It defaults to 1.3 when the version isn't 1.2.

0.22
----

- Added support for svn urls with spaces. Thanks to Brad103 (fixes #537718).

- Updated code and buildout to use newest zc.recipe.egg,
  zc.recipe.testrunner and python-dateutil.

0.21
----

- The admin url is now configured for django 1.1 or higher. Thanks to
  Sam Charrington (fixes #672220).

- Bootstrap.py updated (fixes #501954).

0.20
----

- The recipe know makes the `django` package know to setuptools during install.
  This closes #397864. Thanks to Daniel Bruce and Dan Fairs for the patch.

- Fixed #451065 which fixes a problem with the WSGI log file option.

- Added the posibilty to configure more FCGI related settings. Thanks to Vasily
  Sulatskov for the patch.

0.19.2
------

- The generated WSGI & FCGI scripts are now properly removed when
  options change (fixes #328182). Thanks to Horst Gutmann for the
  patch.

- Scripts are now updated when dependencies change. This fixes #44658,
  thanks to Paul Carduner for the patch.

0.19.1
------

- Applied fix for the change in WSGI script generation. The previous
  release did not work properly.

0.19
----

- When running again with non-newest set the recipe will no longer
  update the Subversion checkout. Thanks to vinilios for the patch.

- The WSGI and FCGI scripts are now generated using Buildout's own
  system. This makes them more similar to the generated manage script
  with regard to the setup of paths. Thanks to Jannis Leidel for the
  patch.

0.18
----

- Paths from eggs and extra-paths now get precedence over the default
  system path (fixes #370420). Thanks to Horst Gutmann for the patch.

- The generated WSGI script now uses the `python` option if
  present. This fixes #361695.

0.17.4
------

- Fixed a problem when not running in verbose mode (fixes #375151).

0.17.3
------

- Removed dependency on setuptools_bzr since it does not seem to work
  like I expected.

0.17.2
------

- Changed the download code to use urllib2. This should make it work
  from behind proxies (fixes #362822). Thanks to pauld for the patch.

0.17.1
------

- Fixed a problem with the new WSGI logging option #348797. Thanks to
  Bertrand Mathieu for the patch.

- Disable generation of the WSGI log if "wsgilog" isn't set, thanks to
  Jacob Kaplan-Moss for the patch.

- Updated buildout.cfg and .bzrignore, thanks Jacob Kaplan-Moss.

0.17
----

- Added an option to specify a log file for output redirection from
  the WSGI script. Thanks to Guido Wesdorp for the patch.

0.16
----

- Subversion aliases are now supported (something like
  svn+mystuff://myjunk). Thanks to Remco for the patch.

0.15.2
------

- Update to move pth-files finder from the __init__ method to the
  install method so it runs in buildout-order, else it looks for pth
  files in dirs that may not yet exist. Thanks to Chris Shenton for
  the update to his original patch.

0.15.1
------

- Update to make the previously added pth-files option better
  documented.

0.15
----

- Added "pth-files" option to add libraries to extra-paths from
  site .pth files. Thanks to Chris Shenton for the patch.

0.14
----

- The recipe now supports creating a FCGI script. Thanks to Jannis
  Leidel for the patch.

- When downloading a Django recipe for the first time the recipe now
  properly reports the url it is downloading from.

0.13
----

- Specifying a user name within a subversion url now works. The code
  that determined the revision has been updated. This fixes issue
  #274004. Thanks to Remco for the patch.

- Updated the template for creating new projects. It now uses the
  current admin system when generating it's `urls.py` file. This fixes
  issue #276255. Thanks to Roland for the patch.

0.12.1
------

- Re-upload since CHANGES.txt was missing from the release

0.12
----

- The recipe no longer executes subversion to determine whether the
  versions is to be downloaded using subversion. This fixes issue
  #271145. Thanks to Kapil Thangavelu for the patch.

- Changed the `pythonpath` option to `extra-paths`. This makes the
  recipe more consistent with other recipes (see issue #270908).

0.11
----

- Another go at fixing the updating problem (#250811) by making sure
  the update method is always called. It would not be called in the
  previous version since the recipe wrote a random secret (if it

src/djangorecipe/recipe.py
from random import choice
import os
import logging
import re

from zc.buildout import UserError
import zc.recipe.egg

from djangorecipe.boilerplate import script_template, versions


class Recipe(object):
    def __init__(self, buildout, name, options):
        # The use of version is deprecated.
        if 'version' in options:
            raise UserError('The version option is deprecated. '
                            'Read about the change on '
                            'http://pypi.python.org/pypi/djangorecipe/0.99')
        self.log = logging.getLogger(name)
        self.egg = zc.recipe.egg.Egg(buildout, options['recipe'], options)

        self.buildout, self.name, self.options = buildout, name, options
        options['location'] = os.path.join(
            buildout['buildout']['parts-directory'], name)
        options['bin-directory'] = buildout['buildout']['bin-directory']

        options.setdefault('project', 'project')
        options.setdefault('settings', 'development')

        options.setdefault('urlconf', options['project'] + '.urls')
        options.setdefault(
            'media_root',
            "os.path.join(os.path.dirname(__file__), 'media')")
        # Set this so the rest of the recipe can expect the values to be
        # there. We need to make sure that both pythonpath and extra-paths are
        # set for BBB.
        if 'extra-paths' in options:
            options['pythonpath'] = options['extra-paths']
        else:
            options.setdefault('extra-paths', options.get('pythonpath', ''))

        options.setdefault('initialization', '')

        # mod_wsgi support script
        options.setdefault('wsgi', 'false')
        options.setdefault('fcgi', 'false')
        options.setdefault('wsgilog', '')
        options.setdefault('logfile', '')

    def install(self):
        base_dir = self.buildout['buildout']['directory']

        project_dir = os.path.join(base_dir, self.options['project'])

        extra_paths = self.get_extra_paths()
        requirements, ws = self.egg.working_set(['djangorecipe'])

        script_paths = []
        # Create the Django management script
        script_paths.extend(self.create_manage_script(extra_paths, ws))

        # Create the test runner
        script_paths.extend(self.create_test_runner(extra_paths, ws))

        # Make the wsgi and fastcgi scripts if enabled
        script_paths.extend(self.make_scripts(extra_paths, ws))

        # Create default settings if we haven't got a project
        # egg specified, and if it doesn't already exist
        if not self.options.get('projectegg'):
            if not os.path.exists(project_dir):
                self.create_project(project_dir)
            else:
                self.log.info(
                    'Skipping creating of project: %(project)s since '
                    'it exists' % self.options)

        return script_paths

    def create_manage_script(self, extra_paths, ws):
        project = self.options.get('projectegg', self.options['project'])

        # Allow to specify settings file outside of project module.
        # Also provide possibility to specify root module by providing dot as
        # first character.
        if '.' in self.options['settings']:
            settings = self.options['settings'].lstrip('.')
        else:
            settings = '%s.%s' % (project, self.options['settings'])

        return zc.buildout.easy_install.scripts(
            [(self.options.get('control-script', self.name),
              'djangorecipe.manage', 'main')],
            ws, self.options['executable'], self.options['bin-directory'],
            extra_paths=extra_paths,
            arguments="'%s'" % settings,
            initialization=self.options['initialization'])

    def create_test_runner(self, extra_paths, working_set):
        apps = self.options.get('test', '').split()
        # Only create the testrunner if the user requests it
        if apps:
            return zc.buildout.easy_install.scripts(
                [(self.options.get('testrunner', 'test'),
                  'djangorecipe.test', 'main')],
                working_set, self.options['executable'],
                self.options['bin-directory'],
                extra_paths=extra_paths,
                arguments="'%s.%s', %s" % (
                    self.options['project'],
                    self.options['settings'],
                    ', '.join(["'%s'" % app for app in apps])))
        else:
            return []

    def create_project(self, project_dir):
        os.makedirs(project_dir)

        # Find the current Django versions in the buildout versions.
        # Assume the newest Django when no version is found.
        version = None
        b_versions = self.buildout.get('versions')
        if b_versions:
            django_version = b_versions.get('django')
            if django_version:
                version_re = re.compile("\d+\.\d+")
                match = version_re.match(django_version)
                version = match and match.group()

        config = versions.get(version, versions['Newest'])

        template_vars = {'secret': self.generate_secret()}
        template_vars.update(self.options)

        self.create_file(
            os.path.join(project_dir, 'development.py'),
            config['development_settings'], template_vars)

        self.create_file(
            os.path.join(project_dir, 'production.py'),
            config['production_settings'], template_vars)

        self.create_file(
            os.path.join(project_dir, 'urls.py'),
            config['urls'], template_vars)

        self.create_file(
            os.path.join(project_dir, 'settings.py'),
            config['settings'], template_vars)

        # Create the media and templates directories for our
        # project
        os.mkdir(os.path.join(project_dir, 'media'))
        os.mkdir(os.path.join(project_dir, 'templates'))

        # Make the settings dir a Python package so that Django
        # can load the settings from it. It will act like the
        # project dir.
        open(os.path.join(project_dir, '__init__.py'), 'w').close()

    def make_scripts(self, extra_paths, ws):
        scripts = []
        _script_template = zc.buildout.easy_install.script_template
        for protocol in ('wsgi', 'fcgi'):
            zc.buildout.easy_install.script_template = \
                zc.buildout.easy_install.script_header + \
                    script_template[protocol]
            if self.options.get(protocol, '').lower() == 'true':
                project = self.options.get('projectegg',
                                           self.options['project'])
                scripts.extend(
                    zc.buildout.easy_install.scripts(
                        [('%s.%s' % (self.options.get('control-script',
                                                      self.name),
                                     protocol),
                          'djangorecipe.%s' % protocol, 'main')],
                        ws,
                        self.options['executable'],
                        self.options['bin-directory'],
                        extra_paths=extra_paths,
                        arguments="'%s.%s', logfile='%s'" % (
                            project, self.options['settings'],
                            self.options.get('logfile'))))
        zc.buildout.easy_install.script_template = _script_template
        return scripts

    def get_extra_paths(self):
        extra_paths = [self.buildout['buildout']['directory']]

        # Add libraries found by a site .pth files to our extra-paths.
        if 'pth-files' in self.options:
            import site
            for pth_file in self.options['pth-files'].splitlines():
                pth_libs = site.addsitedir(pth_file, set())
                if not pth_libs:
                    self.log.warning(
                        "No site *.pth libraries found for pth_file=%s" % (
                         pth_file,))
                else:
                    self.log.info("Adding *.pth libraries=%s" % pth_libs)
                    self.options['extra-paths'] += '\n' + '\n'.join(pth_libs)

        pythonpath = [p.replace('/', os.path.sep) for p in
                      self.options['extra-paths'].splitlines() if p.strip()]

        extra_paths.extend(pythonpath)
        return extra_paths

    def update(self):
        extra_paths = self.get_extra_paths()
        requirements, ws = self.egg.working_set(['djangorecipe'])
        # Create the Django management script
        self.create_manage_script(extra_paths, ws)

        # Create the test runner
        self.create_test_runner(extra_paths, ws)

        # Make the wsgi and fastcgi scripts if enabled
        self.make_scripts(extra_paths, ws)

    def create_file(self, file, template, options):
        if os.path.exists(file):
            return

        f = open(file, 'w')
        f.write(template % options)
        f.close()

    def generate_secret(self):
        chars = 'abcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*(-_=+)'
        return ''.join([choice(chars) for i in range(50)])
